core/src/main/java/org/infinispan/stream/impl/local/EntryStreamSupplier.java

```diff
@@ -44,7 +45,9 @@ public EntryStreamSupplier(Cache<K, V> cache, ConsistentHash hash, Supplier<Stre
       }
       if (segmentsToFilter != null && hash != null) {
          log.tracef("Applying segment filter %s", segmentsToFilter);
-         stream = stream.filter(k -> segmentsToFilter.contains(hash.getSegment(k.getKey())));
+         BitSet bitSet = new BitSet(hash.getNumSegments());
```

Comment:
And LocalCacheStream.filterKeySegments could continue to take a Set and convert it internally to BitSet.